File: test/timeline_test.dart
// The Intelligence Timeline — merged controversy, career, news and
// sentiment events, and the honesty rules around them: a single article
// is not a timeline event, an undated row does not appear, and
// "importance" always reads a real signal rather than inventing one.
import 'package:flutter_test/flutter_test.dart';

import 'package:crititrack/core/domain/models/controversy.dart';
import 'package:crititrack/core/domain/models/media_item.dart';
import 'package:crititrack/core/domain/models/person_facts.dart';
import 'package:crititrack/core/domain/models/sentiment_data.dart';
import 'package:crititrack/core/utils/timeline.dart';

Controversy controversy({
  String title = 'Tax dispute',
  String summary = 'A disagreement over reported income.',
  int severity = 3,
  String status = ControversyStatus.resolved,
  int? year = 2019,
  List<String> sources = const ['https://reuters.com/story'],
}) => Controversy(
  title: title,
  summary: summary,
  category: ControversyCategory.financial,
  severity: severity,
  status: status,
  year: year,
  sources: sources,
);

MediaItem media({
  String id = '1',
  String url = 'https://apnews.com/story',
  String? source = 'AP',
  int? sentimentScore,
  DateTime? publishedAt,
}) => MediaItem(
  id: id,
  type: MediaType.news,
  title: 'Headline',
  url: url,
  source: source,
  sentimentScore: sentimentScore,
  publishedAt: publishedAt ?? DateTime.utc(2026, 3, 12),
);

CareerEntry career({
  int? start = 2018,
  int? end,
  String? role = 'Chief Executive Officer',
  String? organization = 'Firm C',
  String? location,
  String sourceName = 'Wikidata',
  String? sourceUrl = 'https://www.wikidata.org/wiki/Q1',
}) => CareerEntry(
  start: start,
  end: end,
  role: role,
  organization: organization,
  location: location,
  sourceName: sourceName,
  sourceUrl: sourceUrl,
);

List<SentimentSnapshot> trend(List<double> scores) => [
  for (var i = 0; i < scores.length; i++)
    SentimentSnapshot(
      date: '2026-03-0${i + 1}',
      positiveCount: 0,
      negativeCount: 0,
      neutralCount: 0,
      totalMentions: 0,
      dominantEmotion: 'neutral',
      score: scores[i],
    ),
];

List<TimelineEvent> build({
  List<Controversy> controversies = const [],
  List<MediaItem> mediaItems = const [],
  List<CareerEntry> career = const [],
  List<SentimentSnapshot> trend = const [],
}) => buildTimeline(
  controversies: controversies,
  media: mediaItems,
  career: career,
  trend: trend,
);

void main() {
  group('sentimentShiftEvents', () {
    test('emits on a jump past the threshold', () {
      final out = sentimentShiftEvents(trend([50, 65]));
      expect(out.single.change, 15);
      expect(out.single.title, contains('rose sharply'));
    });

    test('ignores a small wobble', () {
      expect(sentimentShiftEvents(trend([50, 54, 51])), isEmpty);
    });
  });

  group('buildTimeline — controversy', () {
    test('places a year-only controversy on Jan 1, flagged approximate', () {
      final e = build(controversies: [controversy()]).single;
      expect(e.date, DateTime.utc(2019, 1, 1));
      expect(e.approxDate, isTrue);
      expect(e.kind, TimelineKind.controversy);
    });

    test("carries the episode's real sources", () {
      final e = build(controversies: [controversy()]).single;
      expect(e.sources.single.label, 'reuters.com');
      expect(e.sources.single.url, 'https://reuters.com/story');
    });

    test('a publication-name-only source has no url', () {
      final e =
          build(
            controversies: [
              controversy(sources: ['Reuters']),
            ],
          ).single;
      expect(e.sources.single.label, 'Reuters');
      expect(e.sources.single.url, isNull);
    });

    test(
      'rates importance from severity and status, not an invented score',
      () {
        final high =
            build(
              controversies: [
                controversy(severity: 5, status: ControversyStatus.ongoing),
              ],
            ).single;
        expect(high.importance, Importance.high);
        expect(high.importanceReason, contains('severity 5/5'));

        final low =
            build(
              controversies: [
                controversy(severity: 1, status: ControversyStatus.historical),
              ],
            ).single;
        expect(low.importance, Importance.low);
      },
    );

    test('drops a controversy with no recorded year', () {
      expect(build(controversies: [controversy(year: null)]), isEmpty);
    });
  });

  group('buildTimeline — career and organization', () {
    test('a role entry becomes a career event', () {
      final e = build(career: [career()]).single;
      expect(e.kind, TimelineKind.career);
      expect(e.title, 'Chief Executive Officer, Firm C');
      expect(e.date, DateTime.utc(2018, 1, 1));
    });

    test('a role-less employer entry becomes an organization event', () {
      final e =
          build(
            career: [
              career(role: null, organization: 'Sidley Austin', start: 1991),
            ],
          ).single;
      expect(e.kind, TimelineKind.organization);
      expect(e.title, 'Sidley Austin');
    });

    test('a leadership title is rated high importance', () {
      final e = build(career: [career(role: 'Chief Executive Officer')]).single;
      expect(e.importance, Importance.high);
    });

    test('an undated career row is dropped', () {
      expect(build(career: [career(start: null)]), isEmpty);
    });
  });

  group('buildTimeline — news grouping', () {
    test('a single article on its own is not a timeline event', () {
      expect(build(mediaItems: [media()]), isEmpty);
    });

    test('two or more sources the same day become one grouped event', () {
      final e =
          build(
            mediaItems: [
              media(id: '1', url: 'https://a.example/1'),
              media(id: '2', url: 'https://b.example/2', source: 'Reuters'),
            ],
          ).single;
      expect(e.kind, TimelineKind.news);
      expect(e.sourceCount, 2);
      expect(e.date, DateTime.utc(2026, 3, 12));
      expect(e.sources, hasLength(2));
    });

    test("averages the grouped items' sentiment, ignoring unscored ones", () {
      final e =
          build(
            mediaItems: [
              media(id: '1', url: 'https://a.example/1', sentimentScore: 80),
              media(id: '2', url: 'https://b.example/2', sentimentScore: 60),
              media(id: '3', url: 'https://c.example/3'),
            ],
          ).single;
      expect(e.sentimentImpact, 70);
    });

    test('does not double-count the same url twice in the source list', () {
      final e =
          build(
            mediaItems: [
              media(id: '1', url: 'https://a.example/1'),
              media(id: '2', url: 'https://a.example/1'),
              media(id: '3', url: 'https://b.example/3'),
            ],
          ).single;
      expect(e.sourceCount, 3);
      expect(e.sources, hasLength(2));
    });

    test('undated media never becomes an event', () {
      MediaItem undated(String id) => MediaItem(
        id: id,
        type: MediaType.news,
        title: 'Headline',
        url: 'https://a.example/$id',
        source: 'AP',
      );
      expect(build(mediaItems: [undated('1'), undated('2')]), isEmpty);
    });
  });

  group('buildTimeline — merge order and relationships', () {
    test('merges every kind, newest first', () {
      final out = build(
        controversies: [controversy(year: 2020)],
        mediaItems: [
          media(id: '1', url: 'https://a.example/1'),
          media(id: '2', url: 'https://b.example/2'),
        ],
        career: [career(start: 2015)],
        trend: trend([50, 50, 70]),
      );
      final kinds = out.map((e) => e.kind).toSet();
      expect(
        kinds,
        containsAll([
          TimelineKind.sentimentShift,
          TimelineKind.news,
          TimelineKind.career,
          TimelineKind.controversy,
        ]),
      );
      // Newest first.
      for (var i = 1; i < out.length; i++) {
        expect(out[i - 1].date.compareTo(out[i].date), greaterThanOrEqualTo(0));
      }
    });

    test('tags nearby events as related without claiming a cause', () {
      final out = build(
        controversies: [controversy(year: 2026)],
        trend: [
          SentimentSnapshot(
            date: '2026-01-01',
            positiveCount: 0,
            negativeCount: 0,
            neutralCount: 0,
            totalMentions: 0,
            dominantEmotion: 'neutral',
            score: 50,
          ),
          SentimentSnapshot(
            date: '2026-01-03',
            positiveCount: 0,
            negativeCount: 0,
            neutralCount: 0,
            totalMentions: 0,
            dominantEmotion: 'neutral',
            score: 20,
          ),
        ],
      );
      final shift = out.firstWhere(
        (e) => e.kind == TimelineKind.sentimentShift,
      );
      expect(shift.relatedTitles, contains('Tax dispute'));
    });

    test('is empty when there is nothing dated at all', () {
      expect(build(), isEmpty);
    });
  });
}

File: site/lib/timeline.test.ts
import { describe, expect, it } from "vitest";
import { buildTimeline, sentimentShiftEvents, withinRangeDays } from "./timeline";
import type { TrendPoint, MediaLink, AttentionPoint } from "./api";
import type { Controversy } from "./controversy";
import type { CareerEntry } from "./career";

const trend = (...s: number[]): TrendPoint[] =>
  s.map((score, i) => ({ date: `2026-03-0${i + 1}`, score, mentions: 0 }));

const controversy = (over: Partial<Controversy> = {}): Controversy => ({
  title: "Tax dispute",
  summary: "A disagreement over reported income.",
  category: "Financial",
  severity: 3,
  status: "resolved",
  year: 2019,
  sources: ["https://reuters.com/story"],
  ...over,
});

const media = (over: Partial<MediaLink> = {}): MediaLink => ({
  id: "1",
  title: "Headline",
  url: "https://apnews.com/story",
  source: "AP",
  type: "news",
  sentimentScore: null,
  publishedAt: "2026-03-12T00:00:00Z",
  ...over,
});

const career = (over: Partial<CareerEntry> = {}): CareerEntry => ({
  start: 2018,
  end: null,
  role: "Chief Executive Officer",
  organization: "Firm C",
  location: null,
  industry: null,
  current: true,
  source: { name: "Wikidata", url: "https://www.wikidata.org/wiki/Q1" },
  ...over,
});

const emptyInput = {
  controversies: [] as Controversy[],
  media: [] as MediaLink[],
  career: [] as CareerEntry[],
  attentionSeries: [] as AttentionPoint[],
  trend: [] as TrendPoint[],
};

describe("withinRangeDays", () => {
  const NOW = new Date("2026-03-15T00:00:00Z").getTime();

  it("null days means no limit", () => {
    expect(withinRangeDays("2010-01-01", null, NOW)).toBe(true);
  });

  it("includes a date inside the window and excludes one outside it", () => {
    expect(withinRangeDays("2026-03-01", 30, NOW)).toBe(true);
    expect(withinRangeDays("2026-01-01", 30, NOW)).toBe(false);
  });

  it("rejects an unparsable date rather than throwing", () => {
    expect(withinRangeDays("not-a-date", 30, NOW)).toBe(false);
  });
});

describe("sentimentShiftEvents", () => {
  it("emits on a jump past the threshold", () => {
    const [e] = sentimentShiftEvents(trend(50, 65));
    expect(e.change).toBe(15);
    expect(e.title).toMatch(/rose sharply/);
  });

  it("ignores a small wobble", () => {
    expect(sentimentShiftEvents(trend(50, 54, 51))).toEqual([]);
  });
});

describe("buildTimeline — controversy", () => {
  it("places a year-only controversy on Jan 1, flagged approximate", () => {
    const [e] = buildTimeline({ ...emptyInput, controversies: [controversy()] });
    expect(e.date).toBe("2019-01-01");
    expect(e.approxDate).toBe(true);
    expect(e.kind).toBe("controversy");
  });

  it("carries the episode's real sources", () => {
    const [e] = buildTimeline({ ...emptyInput, controversies: [controversy()] });
    expect(e.sources).toEqual([{ label: "reuters.com", url: "https://reuters.com/story" }]);
  });

  it("a publication-name-only source has no url", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      controversies: [controversy({ sources: ["Reuters"] })],
    });
    expect(e.sources).toEqual([{ label: "Reuters", url: null }]);
  });

  it("rates importance from severity and status, not an invented score", () => {
    const [high] = buildTimeline({
      ...emptyInput,
      controversies: [controversy({ severity: 5, status: "ongoing" })],
    });
    expect(high.importance).toBe("high");
    expect(high.importanceReason).toMatch(/severity 5\/5/);

    const [low] = buildTimeline({
      ...emptyInput,
      controversies: [controversy({ severity: 1, status: "historical" })],
    });
    expect(low.importance).toBe("low");
  });

  it("drops a controversy with no recorded year", () => {
    expect(
      buildTimeline({ ...emptyInput, controversies: [controversy({ year: undefined })] }),
    ).toEqual([]);
  });
});

describe("buildTimeline — career and organization", () => {
  it("a role entry becomes a career event", () => {
    const [e] = buildTimeline({ ...emptyInput, career: [career()] });
    expect(e.kind).toBe("career");
    expect(e.title).toBe("Chief Executive Officer, Firm C");
    expect(e.date).toBe("2018-01-01");
  });

  it("a role-less employer entry becomes an organization event", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      career: [career({ role: null, organization: "Sidley Austin", start: 1991 })],
    });
    expect(e.kind).toBe("organization");
    expect(e.title).toBe("Sidley Austin");
  });

  it("a leadership title is rated high importance", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      career: [career({ role: "Chief Executive Officer" })],
    });
    expect(e.importance).toBe("high");
  });

  it("an undated career row is dropped", () => {
    expect(
      buildTimeline({ ...emptyInput, career: [career({ start: null })] }),
    ).toEqual([]);
  });
});

describe("buildTimeline — news grouping", () => {
  it("a single article on its own is not a timeline event", () => {
    expect(
      buildTimeline({ ...emptyInput, media: [media()] }),
    ).toEqual([]);
  });

  it("two or more sources the same day become one grouped event", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      media: [
        media({ id: "1", url: "https://a.example/1" }),
        media({ id: "2", url: "https://b.example/2", source: "Reuters" }),
      ],
    });
    expect(e.kind).toBe("news");
    expect(e.sourceCount).toBe(2);
    expect(e.date).toBe("2026-03-12");
    expect(e.sources).toHaveLength(2);
  });

  it("averages the grouped items' sentiment, ignoring unscored ones", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      media: [
        media({ id: "1", url: "https://a.example/1", sentimentScore: 80 }),
        media({ id: "2", url: "https://b.example/2", sentimentScore: 60 }),
        media({ id: "3", url: "https://c.example/3", sentimentScore: null }),
      ],
    });
    expect(e.sentimentImpact).toBe(70);
  });

  it("does not double-count the same url twice in the source list", () => {
    const [e] = buildTimeline({
      ...emptyInput,
      media: [
        media({ id: "1", url: "https://a.example/1" }),
        media({ id: "2", url: "https://a.example/1" }),
        media({ id: "3", url: "https://b.example/3" }),
      ],
    });
    expect(e.sourceCount).toBe(3);
    expect(e.sources).toHaveLength(2);
  });

  it("undated media (no publishedAt) never becomes an event", () => {
    expect(
      buildTimeline({
        ...emptyInput,
        media: [
          media({ id: "1", publishedAt: undefined }),
          media({ id: "2", publishedAt: undefined }),
        ],
      }),
    ).toEqual([]);
  });
});

describe("buildTimeline — attention and sentiment (existing behaviour preserved)", () => {
  const spikeSeries = (): AttentionPoint[] => [
    { date: "2026-01-01", views: 100 },
    { date: "2026-01-02", views: 110 },
    { date: "2026-01-03", views: 105 },
    { date: "2026-01-04", views: 95 },
    { date: "2026-01-05", views: 100 },
    { date: "2026-01-06", views: 5000 },
  ];

  it("flags a genuine spike day and carries the raw view count", () => {
    const out = buildTimeline({ ...emptyInput, attentionSeries: spikeSeries() });
    expect(out).toHaveLength(1);
    expect(out[0].kind).toBe("attention-spike");
    expect(out[0].attentionImpact).toBe(5000);
  });

  it("a sentiment shift still appears and carries its delta", () => {
    const out = buildTimeline({ ...emptyInput, trend: trend(50, 50, 70) });
    expect(out).toHaveLength(1);
    expect(out[0].kind).toBe("sentiment-shift");
    expect(out[0].change).toBe(20);
  });
});

describe("buildTimeline — merge order and relationships", () => {
  it("merges every kind, newest first", () => {
    const out = buildTimeline({
      controversies: [controversy({ year: 2020 })],
      media: [
        media({ id: "1", url: "https://a.example/1", publishedAt: "2026-03-02T00:00:00Z" }),
        media({ id: "2", url: "https://b.example/2", publishedAt: "2026-03-02T00:00:00Z" }),
      ],
      career: [career({ start: 2015 })],
      attentionSeries: [],
      trend: trend(50, 50, 70),
    });
    const kinds = out.map((e) => e.kind);
    expect(kinds).toContain("sentiment-shift");
    expect(kinds).toContain("news");
    expect(kinds).toContain("career");
    expect(kinds).toContain("controversy");
    // Newest first.
    expect(out[0].date >= out[out.length - 1].date).toBe(true);
  });

  it("tags nearby events as related without claiming a cause", () => {
    const out = buildTimeline({
      ...emptyInput,
      controversies: [controversy({ year: 2026 })],
      trend: [
        { date: "2026-01-01", score: 50, mentions: 0 },
        { date: "2026-01-03", score: 20, mentions: 0 },
      ],
    });
    const sentiment = out.find((e) => e.kind === "sentiment-shift")!;
    expect(sentiment.relatedTitles).toContain("Tax dispute");
  });

  it("is empty when there is nothing dated at all", () => {
    expect(buildTimeline(emptyInput)).toEqual([]);
  });
});

describe("buildTimeline — Step 16 Change Detection integration", () => {
  const change = (over: Partial<import("./changes").ChangeEvent> = {}): import("./changes").ChangeEvent => ({
    changeId: "c1",
    entityId: "x",
    changeType: "CRITISCORE_CHANGE",
    severity: "MAJOR",
    title: "CritiScore increased +13",
    summary: "reason",
    previousValue: "48",
    currentValue: "61",
    detectedAt: "2026-09-05T00:00:00Z",
    effectiveDate: null,
    evidenceIds: [],
    relatedClaimIds: [],
    methodologyVersion: "2.0",
    confidence: "HIGH",
    sourceCoverage: null,
    ...over,
  });

  it("folds a CritiScore/claim/coverage/profile change in as its own 'change' event", () => {
    const out = buildTimeline({ ...emptyInput, changeEvents: [change()] });
    expect(out).toHaveLength(1);
    expect(out[0].kind).toBe("change");
    expect(out[0].title).toBe("CritiScore increased +13");
    expect(out[0].importance).toBe("high");
  });

  it("never duplicates a controversy/career/news/sentiment change already derivable from the same data", () => {
    const out = buildTimeline({
      ...emptyInput,
      controversies: [controversy({ year: 2026 })],
      changeEvents: [
        change({ changeType: "CONTROVERSY_CHANGE", title: "New supported controversy: Tax dispute" }),
      ],
    });
    // Only the controversy's own timeline entry — the duplicate-shaped
    // CONTROVERSY_CHANGE is not one of the folded-in types.
    expect(out).toHaveLength(1);
    expect(out[0].kind).toBe("controversy");
  });

  it("with no changeEvents supplied, behaves exactly as before Step 16", () => {
    const out = buildTimeline({ ...emptyInput, controversies: [controversy({ year: 2026 })] });
    expect(out).toHaveLength(1);
    expect(out.every((e) => e.kind !== "change")).toBe(true);
  });

  it("uses the change's effectiveDate when known, else falls back to detectedAt", () => {
    const withDate = buildTimeline({
      ...emptyInput,
      changeEvents: [change({ effectiveDate: "2020" })],
    });
    expect(withDate[0].date).toBe("2020");
    expect(withDate[0].approxDate).toBe(false);

    const withoutDate = buildTimeline({ ...emptyInput, changeEvents: [change({ effectiveDate: null })] });
    expect(withoutDate[0].date).toBe("2026-09-05");
    expect(withoutDate[0].approxDate).toBe(true);
  });
});
